### ISSUE TITLE
Trace function for debug output

### ISSUE THREAD

**Issue by laurmaedje** [typst/typst#1669]
### Description

Debug output via a `trace` function or something similar. 

I don't want to encourage lots of logging as I want to prevent the situation where libraries dump screens of stuff at you (bad from a UX perspective). It's also a performance concern because tracing would use tracked memoization, which has some overhead.

Being able to filter traces would also make it easier to extract some specific data output without having other stuff interrupt it. We'd also need to decide where to log to (probably just stdout? some people wanted the typst -> PDF via stdin -> stdout so that might conflict).

### Use Case

- Debuggingg
- Extracting information from the document, though this would long-term be better served by a query API (see #1056)

**Comment by user202729** [comment]
Remark: in LaTeX, the log output is "unstructured", which leads to situation such as latexmk greps for `No file` or `Rerun to get` in the log file to extract content, or log parser grep for `! ` at the start of file.

It's preferable to somehow make the log file "easy to parse" so complicated/unreliable regex solutions is not needed.

I don't see a specific way to solve it so far, but it would be a good idea to take that into consideration.

**Comment by imbrem** [comment]
I dislike the stdout idea since it leads down a path littered with, as my professor put it, "TeX droppings." A log file is preferable: only serious errors and warnings should be in stdout/stderr imo

**Comment by bluebear94** [comment]
Perhaps we could require a flag to be passed in order to print log messages. However, this still has the risk of people “abusing the software” in the way @user202729 described.

**Comment by user202729** [comment]
> I want to prevent the situation where libraries dump screens of stuff at you

I think the situation is prevented in Python precisely because it has proper logging facility (`logging` module, but normally the log by `logging.info` is not visible to the user unless they do something to enable it)

------

A reasonable "goalpost" would be a real programming language, I believe. 

Personally, for debugging purpose, I use
* C++: gdb (step by step, print out values)
* Python: interactive shell/notebook (you can "sort of" step by step this one, and also it's very convenient to edit some lines of the code and print out values without re-computing everything from the beginning) for small scripts, and logging for larger things.

It's still very inconvenient to have to (edit a line in the program, switch to terminal, type compilation command, switch back, edit another line, etc.) however, but with `typst watch`-equivalent it should not be too bad modulo terminal scrolling. For other languages I have similar facility.

and for interfacing with other programs, output to a file is a reasonable option -- although
* for a real programming language, things like shared-memory or message-passing would be _more efficient_ without the overhead of the file system (this may require a programming API for the target language however)
* avoid outputting to a single file -- different "packages" (whatever it's called here) would prefer to have different auxiliary files in order to not need to "regex"-parse out each other's output.

**Comment by laurmaedje** [comment]
For more convenient inspection of values there is already the IDE integration with hover tooltips. The tracing would be specifically for people that want to work in the terminal or display multiple values at once.

We must be careful that anything we do here is compatible with memoization. Shared memory and message passing seem problematic in this regard. The proposed tracing couldn't directly print things because that's a side effect. Instead all logs would be collected in a memoized-aware way and printed at once.

**Comment by lf-** [comment]
The use case I want a trace function for personally is the case when I have a bad error message from putting e.g. a dictionary into an argument of `table()`. Currently I can get useful debug output by either dumping `repr` into the output file (but this sucks to read due to non monospace font so I get wacked by dyslexia on the indents) or dumping the offending value into `panic` but I have to have some idea of the conditions where the value is bad.

**Comment by Andrew15-5** [comment]
> I don't want to encourage lots of logging as I want to prevent the situation where libraries dump screens of stuff at you

https://github.com/typst/typst/issues/2137#issue-1895331630:

> So I was wondering if we can get an easy `print`/`println`/`log` function. And maybe only enable them when a special option/flag is used in the CLI (`--debug` or use the `-v`).

With an opt-out by default "thing"/"state" there won't be any logs anywhere. Maybe try to create a "simple" printing function and then make it more robust and things like that. The earlier we can debug our code, the better. The project is still in beta anyway.

**Comment by Andrew15-5** [comment]
At first, I was thinking about `warning` function that would be shown exactly like the error function is shown, but it won't be an error. But the goal is not really to warn someone (although such function would be handy in some cases), therefore it should be called like `notice`/`note` (I think I got that from the PL/pgSQL). Maybe it could be a separate function or a boolean option: "just print stuff" or "print stuff and include the location from where the stuff was printed" (path to the file and line number).

**Comment by nikhilweee** [comment]
Eagerly waiting for this functionality. Just started using using functions and I'm already having a hard time debugging them.

**Comment by laurmaedje** [comment]
@nikhilweee If you don't already use them, I suggest trying the hover tooltips in the web app or Typst LSP for debugging. A logging function would still be useful of course, but they go a long way to understanding the runtime behaviour.

**Comment by zmaril** [comment]
Hello, I have been enjoying typst very much the last few weeks. I find myself very surprised though that there is no way of printing out or otherwise debugging typst scripts that I am writing. I understand that there is the LSP, however for anything with a nested call stack or a recursive function, that isn't viable. I would very much like a way of printing out to stdout whenever I need to and it's very limiting to not have one.

**Comment by SolidTux** [comment]
A warning would definitely be useful. For example, I'm currently writing a template that does some automatic layouting and it would be nice to give a warning in case the leading or other values reach some uncommon values that are probably unintended.

**Comment by AdrianVollmer** [comment]
+1 for warnings. LaTeX does warnings (like the dreaded overfull box or missing references) and I found it very helpful. I guess in Typst everything is an error, but maybe there could also be an option to treat warnings as errors? Or the other way round, like some sort of draft mode?

**Comment by inferiorhumanorgans** [comment]
My two cents:  Simply adding the log crate as a dependency to the typst libraries would be a big step forward and be completely agnostic to whatever decision gets made for typst-cli.  If a decision is made to use tracing for profiling in the future, that can be done with little-to-no effort.  For typst-cli my vote is log to `stderr` and not `stdout` with the option to log to a file.

What I've used for my projects is either `log` + `env_logger` or `tracing` + `EnvFilter`.  Both stacks support the five common severities (trace, debug, info, warn, error) and appropriate use of these would avoid overwhelming users.  Depending on the expected verbosity I've then either used the feature flags to turn the less severe levels into no-ops or initialized them with different default levels depending on the build config (e.g. debug vs release).

The environment variable based filters allow users to override the default level by setting the `RUST_LOG` env var, and they both allow for more fine grained logging (per module or function depending).  This could also be comingled with or replaced by a command line argument.

**Comment by japgolly** [comment]
We already have a `panic` function, and typst already outputs warnings when it needs to. Could we not add a `warn` function that simply issues warnings the same way the internal code does. It would compliment `panic()` nicely too

**Comment by tovrstra** [comment]
The following suggestion may address the point raised by @user202729. An option to format all output (including the envisioned logging output) as [new-line delimited JSON](https://en.wikipedia.org/wiki/JSON_streaming#Newline-delimited_JSON) would be helpful. This is [supported by many existing tools](https://en.wikipedia.org/wiki/JSON_streaming#Newline-delimited_JSON_2). Having a first line as follows could then help with resolving compatibility issues:

```json
{"typst-version": "v0.13", "typst-log-schema-version": "1"}
```

This could be useful for external tools that want to pick up error or warning messages and act upon them. For example, I'm currently regexing the `typst compile` output to detect any missing files, so they can be built when needed. (It seemed practically impossible to extract that information from the typst document upfront. I'd like to be proven wrong, though.)

Along those lines, It would be useful to log all files written and read. I'm currently relying on `--make-deps` to extract this information, but this file is only written when the build succeeds. (That is by itself defendable in the context of tools like make, but it makes it hard to build good workflows in general.)

**Comment by Jollywatt** [comment]
I strongly agree that adding `warn()` or `log()` functions will lead to annoying over-use.

However, when developing a complex package, it can be extremely difficult to debug with only `panic()`. It can be hard or even impossible to avoid panicking and halting prematurely.

An escape hatch, even if it is very well hidden, would be a life-saver. For example, we could expose a function `sys.log()` which does not exist unless typst is invoked with `--enable-debug-logging`, which does basic (memoisation-unaware) logging to stdout or a file.

If I could, I would have forked typst and added something quick-and-dirty like this in order make debugging fletcher easier!

**Comment by cskeeters** [comment]
Wait, I missed this too.  How does one use the Sink/Traced values?  Can this work with Neovim and tinymist preview?  Where does the output go?

**Comment by laurmaedje** [comment]
In the web app or if your editor supports LSP, you can view the values by hovering over a variable. Not sure what the neovim situation is.

**Comment by Andrew15-5** [comment]
In a sane LSP configuration for Neovim, it's a `<s-k>` keymap or `K`. Though in the past I remember that it didn't work always, but nowadays, it still works when I need it. Only works when the piece of code that was executed, so it doesn't work for dead code.

**Comment by Andrew15-5** [comment]
The (unrelated) Neovim questions are out of scope for this issue, so please ask them in https://forum.typst.app. AFAIK, tinymist just adds the value of a certain variable/function output to the docs window, so usually instead of, or in addition to the generated docs it will also show the result value. I'm not actually sure if it already keeps all the values ready to show, or if it indeed recompiles the document to get them.

**Comment by Andrew15-5** [comment]
This issue has been mentioned on **Typst Forum**. There might be relevant details there:

https://forum.typst.app/t/why-isnt-there-a-function-to-output-to-stdout/3362/2

**Comment by AprilGrimoire** [comment]
> In the web app or if your editor supports LSP, you can view the values by hovering over a variable. Not sure what the neovim situation is.

That doesn't work in some use cases, for example when trying to debug a custom function.

**Comment by johannesrave** [comment]
i am using typst to build a CLI-tool (or a set of tools rather) for my colleagues to use.
i would've liked to print for debugging as well as give users helpful status and error messages, such as "please add a portrait-image of yourself to the img folder". 

i CAN try to think of all potential preconditions regarding existing files etc and move that checking-logic into a bash-script, however co-locating the error-conditions and the messages with the actual rendering logic would be a LOT more convenient and less error-prone.

being a developer i also use the default logging severities in my daily work and would appreciate being able to use something like that (probably wouldn't need five levels but maybe info, warning, error?)

**Comment by lordkekz** [comment]
TL;DR +1 for json line logging, hidden behind a CLI option. **I also suggest ignore directives and disabling debug logging for published packages.**

As a package author, I have experienced situations in which I would like to warn about a very likely unwanted situation. Think "overfull hbox". As a package user, I also want to be able to look for such non-panic-worthy problems, but I don't want to be flooded in them every time I build my document.

I think this can be solved by adding a way to generate structured warnings from Typst code, which can be exported or queried, but which do not appear in the stdout by default. I would hope for such warnings to be displayed inline, like panic-messages already are when using e.g. Tinymist LSP.

I would also like to be able to dismiss warnings in code, similar to e.g. [shellcheck's ignore directive](https://github.com/koalaman/shellcheck/wiki/Ignore#ignoring-errors). If I really want that overfull hbox, the warning should not clutter among others. If a certain packages generates too many warnings I don't want to see, I should be able to disable the particular warning type or all warnings from that package. The identifiers for particular could be chosen like `<package-name>:<warning-label>` and IMO it may be acceptable if multiple warnings have the same label (i.e. the same kind of misconfiguration might occur in multiple places of package code, which needn't be exposed to the user).

Additionally, published Typst packages could have their debug logging disabled either as a compiler feature or a requirement for publishing. This would ensure that only user-targeted logging remains.

**Comment by sspencer-arine** [comment]
I have a situation where we need to know what actually went into a PDF and we can't always know or expect the input context to be fully satisfied.

Having a logging function or output collection of some kind would allow for enough information to gauge this.

The alternative for my use case would be to generate typst documents through a template language rather than leverage more powerful features.

**Comment by tovrstra** [comment]
> I have a situation where we need to know what actually went into a PDF

The `--deps` option may be of interest. (Not entirely sure what you need to infer exactly.)

**Comment by omentic** [comment]
For those dropping by: I have been using the following terrible hack (thanks, @cady-b!)

```typst
#let warn(body) = {
  let my-message = [#(label(repr(body)))]
}

#warn("foo")
```

(note that it is a terrible hack and only approximately works. but it approximately works.)
